### ISSUE TITLE
"Can't find variable: self"

### ISSUE THREAD

**Issue by alexanderkjeldaas** [GoogleChrome/proxy-polyfill#26]
Just got this with 1.7.0 when used in react-native.

Don't have too much info as I've downgraded to 1.6.0.

**Comment by samthor** [comment]
[`self` is used now](https://github.com/GoogleChrome/proxy-polyfill/pull/19) as it should equal `window` or whatever your global scope is. I'm confused, because node or other runtimes (which admittedly, don't have `self`) should run this line-

```js
("undefined"!==typeof process&&"[object process]"=={}.toString.call(process)?global:self);
```

which shouldn't even run `self` because process/global fires first.

What environment/etc does react-native run in? Is there a repro case?

**Comment by m-vdb** [comment]
@samthor the second condition fails. On react-native, `{}.toString.call(process) == "[object Object]"`. Hope that helps :)

**Comment by samthor** [comment]
Does react-native use 'global' then?

**Comment by m-vdb** [comment]
yes `global` exists, but not `self`, and since the condition in the ternary doesn't return true, it resolves to `self`.

**Comment by hojberg** [comment]
Anyone find an alternative to this? Running into it as well..

**Comment by m-vdb** [comment]
pinning version to `1.6.0` fixes it

**Comment by tslater** [comment]
@m-vdb version 1.6.0 of proxy-polyfill? Or what?

**Comment by m-vdb** [comment]
oups, I mistyped, it's version `0.1.6` of `proxy-polyfill`

**Comment by nickleach** [comment]
So are there any plans to fix this without downgrading?

**Comment by fxfactorial** [comment]
@samthor Ping if some attention can be put on this please

**Comment by samthor** [comment]
I'd happily accept PRs that help update the environment detection mode.

Alternatively, we now ship the code 'on its own' without the export code, under `src/proxy.js`, as a commonJS module. Could you just import it directly?

**Comment by machour** [comment]
@samthor just got hit by this bug on React-Native. Importing `src/proxy.js` doesn't work for me.

It seems that this problem is only affecting React Native, where the result is "[object Object]" instead of "[object process]"

After some digging, I came across https://github.com/facebook/react-native/commit/3c65e62183ce05893be0822da217cb803b121c61. So `navigator.product === "ReactNative"` in that environment.

Changing the test to the following works for me:

```
(function(scope) {
  if (scope['Proxy']) {
    return;
  }
  scope.Proxy = require('./proxy.js')();
  scope.Proxy['revocable'] = scope.Proxy.revocable;
})(
  ('undefined' !== typeof process &&
    '[object process]' === {}.toString.call(process)) ||
  ('undefined' !== typeof navigator && navigator.product === 'ReactNative')
    ? global
    : self
);

```
